docs/content/preview/api/ysql/user-defined-subprograms-and-anon-blocks/language-plpgsql-subprograms/plpgsql-syntax-and-semantics/executable-section/loop-exit-continue-statements/query-for-loop.md

```diff
@@ -0,0 +1,370 @@
+---
+title: Query for loop [YSQL]
+headerTitle: The "query for loop"
+linkTitle: Query for loop
+description: Describes the syntax and semantics of the "query for loop" [YSQL]
+menu:
+  preview:
+    identifier: query-for-loop
+    parent: loop-exit-continue-statements
+    weight: 40
+type: docs
+showRightNav: true
+---
+
+The _query for loop_ has three variants according to how the query is defined:
+
+- with a subquery
+- with a so-called _bound ref cursor variable_
+- with a dynamic _SQL_ statement
+
+## "Query for loop" with a subquery
+
+Do this:
+
+```plpgsql
+\c :db :u
+drop schema if exists s cascade;
+create schema s;
+create table s.t(k serial primary key, v int not null);
+insert into s.t(v) select generate_series(0, 24, 5);
+
+create function s.f(k_lo in int, k_hi in int)
+  returns table(z text)
+  set search_path = pg_catalog, pg_temp
+  language plpgsql
+as $body$
+declare
+  rec record;
+begin
+  for rec in (
+    select k, v
+    from s.t
+    order by k)
+  loop
+    z := rpad(rec.k::text, 10)||rec.v::text;      return next;
+  end loop;
+  z := '';                                        return next;
+  for rec in (
+    select v
+    from s.t
+    where k between k_lo and k_hi
+    order by k)
+  loop
+    z := rec.v::text;                             return next;
+  end loop;
+end;
+$body$;
+```
+
+Test it thus:
+
+```plpgsql
+select s.f(2, 4);
+```
+
+This is the result:
+
+```plpgsql
+ 1         0         
+ 2         5         
+ 3         10        
+ 4         15        
+ 5         20        
+ 
+ 5
+ 10
+ 15
+```
+
+Look at the first _query for loop_:
+
+```plpgsql
+for rec in (
+  select k, v
+  from s.t
+  order by k)
+loop
+  z := rpad(rec.k::text, 10)||rec.v::text;        return next;
+end loop;
+```
+
+Notice these points:
+
+- The _loop variable_ for a _query for loop_ must be explicitly declared.
+- The _loop variable_, _rec_, has the data type _record_. This is a so-called pseudo-type. While you _can_ use it as the data type of a subprogram's formal argument or of a PL/pgSQL variable, you cannot use it as the datatype of, for example, a table's column. The attempt _"create table s.x(k serial primary key, v record)"_ causes the _42P16_ error:
```

Comment:
I did a global search for _datatype_ in the ordinary prose. I found and fixed very many like you pointed out here (and including that one). It's tricky because we have a directory called `datatypes`. (It should have been called `data-types` but that choice was made before I started at YB. The consequence is that scores of relative URLs are spelled with `datatype`. It was be possible, but time-consuming and error-prone to fix this now.

I used this regular expression search: `[^/#_\-]datatype`. but, even so, I got lots of false positives because, believe it or not, it's a PG keyword in some use cases and we necessarily inherit it. Grrrr...

**Fixed.**